### ISSUE TITLE
Unable to approve translation to own Kata

### ISSUE THREAD

**Issue by DonaldKellett** [codewars/codewars.com#2096]
[@chunes](https://www.codewars.com/users/5e85e453b633df00150196e1) recently made a Factor translation to my [Convert a linked list to a string](https://www.codewars.com/kata/582c297e56373f0426000098/) Kata, but when I click the "Approve" button, nothing happens. Upon inspecting the console, I found an "unhandled promise rejection" error:

![螢幕截圖 2020-04-14 22 14 39](https://user-images.githubusercontent.com/7777884/79235117-c6e3bb00-7e9d-11ea-80b5-7b803c0e250e.png)

- OS: macOS Catalina 10.15.4
- Username: [@donaldsebleung](https://www.codewars.com/users/donaldsebleung)
- Browser: Safari 13.1

**Comment by DonaldKellett** [comment]
Another example: this time it's not my Kata but a week-old translation ([@shingtaklam1324](https://www.codewars.com/users/5a200c8e45ab038a0a000132)'s [Lean translation](https://www.codewars.com/kumite/5e8c9e5a65e6f40020772658?sel=5e8c9e5a65e6f40020772658) to [Introduction to bijection, cardinality and infinite sets](https://www.codewars.com/kata/5ca3fcc8f7fb0800201ba88f/)) which ought to be approvable by me since I have 6k+ honor. Again, after clicking on "Approve" and confirming my decision, nothing happens, but there is an unhandled promise rejection error in the console:

![螢幕截圖 2020-04-14 22 22 19](https://user-images.githubusercontent.com/7777884/79235916-d1528480-7e9e-11ea-91e3-f5487cd47d46.png)

**Comment by DonaldKellett** [comment]
> [@chunes](https://www.codewars.com/users/5e85e453b633df00150196e1) recently made a Factor translation to my [Convert a linked list to a string](https://www.codewars.com/kata/582c297e56373f0426000098/) Kata, but when I click the "Approve" button, nothing happens. Upon inspecting the console, I found an "unhandled promise rejection" error:
> 
> ![螢幕截圖 2020-04-14 22 14 39](https://user-images.githubusercontent.com/7777884/79235117-c6e3bb00-7e9d-11ea-80b5-7b803c0e250e.png)
> 
> * OS: macOS Catalina 10.15.4
> * Username: [@donaldsebleung](https://www.codewars.com/users/donaldsebleung)
> * Browser: Safari 13.1

I've now checked and this Factor translation has been added but the translation approval process seems to take a while and there is no indication in the UI whatsoever that it's loading / processing the request.

**Comment by DonaldKellett** [comment]
About the week-old translation: I've just tried approving it again and found another error in the console (413 Request Entity Too Large) which may be of interest:

![螢幕截圖 2020-04-14 22 30 03](https://user-images.githubusercontent.com/7777884/79236492-97ce4900-7e9f-11ea-8484-2f3c4cfc85c0.png)

**Comment by Steffan153** [comment]
I don't know if this will help, but does it work in Chrome?

**Comment by Steffan153** [comment]
I just tried to approve the Lean translation in Chrome, it didn't work.
![image](https://user-images.githubusercontent.com/40404519/79239728-22766000-7e2e-11ea-84bf-763ba4fe36b7.png)

I recently had a problem like this with Blind4Basics' translation, and when I rejected it and he forked it again, it worked. Can you try that?

**Comment by FArekkusu** [comment]
I approved/rejected some translations today; no progress bar was shown, and the page content didn't update, but the approval/rejection worked normally.

**Comment by Steffan153** [comment]
Wut? And I just approved the lean translation. Huh. It only showed when I refreshed the page tho.

**Comment by kazk** [comment]
I think this was reported before.

The problem is that the translation approval button enqueues a background job to copy the translation code in and do other stuff. So successful response just means it was enqueued and doesn't mean it's done. It can also take some time depending on the job queue status. If you get `401` (unauthorized), that means you can't approve it.

The entire translation workflow needs to be improved.

**Comment by kazk** [comment]
I'm not sure what happened for the first one with 500 yet.

**Comment by kazk** [comment]
For https://github.com/codewars/codewars.com/issues/2096#issuecomment-613477642, `413 Request Entity Too Large` comes from Sentry (error tracking service). That just means we failed to record the error because we sent too much to them and it's not related to the original error.

**Comment by kazk** [comment]
For https://github.com/codewars/codewars.com/issues/2096#issuecomment-613474594, I don't think it was a week old when you tried. It was published at `2020-04-07T15:38:43.000Z` and your comment was posted on `2020-04-14T00:25:00.000Z`. So you had to wait few more hours.

I need to fix this caching issue so those buttons only show when you can actually approve.

**Comment by kazk** [comment]
Related: https://github.com/codewars/codewars.com/issues/1793#issuecomment-482893334

**Comment by Steffan153** [comment]
> If you get 401 (unauthorized), that means you can't approve it.

In B4B's case, I had perfect permission. Again, the translation was a few months old already, and I have 4k honor. It gave me 401 when I clicked approve, but finally I rejected it and he forked it, and it worked.

Also, the first time I tried to approve the Lean translation, it _didn't_ work, but later it worked.

**Comment by kazk** [comment]
Doesn't it say "Last Week"? You can hover over it to see the exact timestamp in UTC.

Anyways, this confusion comes from the buttons being shown even when you can't approve it. So I'm looking into the caching issue that's preventing me from showing it based on the status.

**Comment by Steffan153** [comment]
Okay, I was misunderstanding, last week doesn't mean 1 week, 7 days precisely.

> Anyways, this confusion comes from the buttons being shown even when you can't approve it.

But when a user has the power to approve translations, it might be nice to show _why_ they can't approve it, just for info.

Do you have any idea what happened with B4B's case? Here's the translation:https://www.codewars.com/kumite/5c08186e85e9ff8200000251?sel=5e747d98762315001ebbc9c5

**Comment by kazk** [comment]
> But when a user has the power to approve translations, it might be nice to show why they can't approve it, just for info.

Yeah, it should be easy to add that once I know WTF is wrong with this caching.

> Do you have any idea what happened with B4B's case?

You got `401` so the background job wasn't even enqueued. It should be some permission issue, but I don't know the exact cause.

Kumite/Translation is really messy.

**Comment by Steffan153** [comment]
I think this is basically fixed now, I even see appropriate messages.

**Comment by kazk** [comment]
Yeah, I started to fix some of the error handling on frontend. I still haven't figured out the caching issue though.

I've also changed to check the language version of the translation _before_ queuing the job to merge into the kata. I'm guessing translations having invalid language version was one of the causes for https://github.com/codewars/codewars.com/issues/1793#issuecomment-482893334 .
